chips/rp2040/src/pio.rs

```diff
@@ -0,0 +1,1233 @@
+pub enum PIONumber {
+    PIO0 = 0,
+    PIO1 = 1,
+}
+
+impl RPGpio {
+    fn from_u32(value: u32) -> RPGpio {
+        match value {
+            0 => RPGpio::GPIO0,
+            1 => RPGpio::GPIO1,
+            2 => RPGpio::GPIO2,
+            3 => RPGpio::GPIO3,
+            4 => RPGpio::GPIO4,
+            5 => RPGpio::GPIO5,
+            6 => RPGpio::GPIO6,
+            7 => RPGpio::GPIO7,
+            8 => RPGpio::GPIO8,
+            9 => RPGpio::GPIO9,
+            10 => RPGpio::GPIO10,
+            11 => RPGpio::GPIO11,
+            12 => RPGpio::GPIO12,
+            13 => RPGpio::GPIO13,
+            14 => RPGpio::GPIO14,
+            15 => RPGpio::GPIO15,
+            16 => RPGpio::GPIO16,
+            17 => RPGpio::GPIO17,
+            18 => RPGpio::GPIO18,
+            19 => RPGpio::GPIO19,
+            20 => RPGpio::GPIO20,
+            21 => RPGpio::GPIO21,
+            22 => RPGpio::GPIO22,
+            23 => RPGpio::GPIO23,
+            24 => RPGpio::GPIO24,
+            25 => RPGpio::GPIO25,
+            26 => RPGpio::GPIO26,
+            27 => RPGpio::GPIO27,
+            28 => RPGpio::GPIO28,
+            29 => RPGpio::GPIO29,
+            _ => panic!(
+                "Unknown value for GPIO pin: {} (should be from 0 to 29)",
+                value
+            ),
+        }
+    }
+}
+
+// The FIFO queues can be joined together for twice the length in one direction.
+#[derive(PartialEq)]
+pub enum PioFifoJoin {
+    PioFifoJoinNone = 0,
+    PioFifoJoinTx = 1,
+    PioFifoJoinRx = 2,
+}
+
+const STATE_MACHINE_NUMBERS: [SMNumber; NUMBER_STATE_MACHINES] =
+    [SMNumber::SM0, SMNumber::SM1, SMNumber::SM2, SMNumber::SM3];
+
+pub struct Pio {
+    registers: StaticRef<PioRegisters>,
+    pio_number: PIONumber,
+    xor_registers: StaticRef<PioRegisters>,
+    set_registers: StaticRef<PioRegisters>,
+    _clear_registers: StaticRef<PioRegisters>,
+}
+
+// 'MOV STATUS' types.
+#[derive(Clone, Copy)]
+pub enum PioMovStatusType {
+    StatusTxLessthan = 0,
+    StatusRxLessthan = 1,
+}
+
+/// PIO State Machine configuration structure
+///
+/// Used to initialize a PIO with all of its state machines.
+pub struct StateMachineConfiguration {
+    pub out_pins_count: u32,
+    pub out_pins_base: u32,
+    pub set_pins_count: u32,
+    pub set_pins_base: u32,
+    pub in_pins_base: u32,
+    pub side_set_base: u32,
+    pub side_set_opt_enable: bool,
+    pub side_set_bit_count: u32,
+    pub side_set_pindirs: bool,
+    pub wrap: u32,
+    pub wrap_to: u32,
+    pub in_shift_direction_right: bool,
+    pub in_autopush: bool,
+    pub in_push_threshold: u32,
+    pub out_shift_direction_right: bool,
+    pub out_autopull: bool,
+    pub out_pull_threshold: u32,
+    pub jmp_pin: u32,
+    pub out_special_sticky: bool,
+    pub out_special_has_enable_pin: bool,
+    pub out_special_enable_pin_index: u32,
+    pub mov_status_sel: PioMovStatusType,
+    pub mov_status_n: u32,
+    pub div_int: u32,
+    pub div_frac: u32,
+}
```

Comment:
I am not too familiar with how PIO is used in practice, but am a bit surprised that you expect all the individual fields of this struct to be accessible outside of this crate.

Reply:
Honestly not sure, was thinking that if an user wanted to make a program in another folder or just in main.rs for example they might need to change just a few fields and leave the others as default - but to use them they need to be pub. Any ideas are welcome, I really don't know if this is the right approach.

Reply:
Fair, if we expect the state machine logic to live outside this file it might be hard to predict which fields need to be accessible. Fine to leave this as-is